File: python/year_2015/day_02.py
"""
https://adventofcode.com/2015/day/2
"""


from pathlib import Path
def process(text):
    return (sorted(map(int, line.split("x"))) for line in text.split())


def part_1(text):
    return sum(3 * a * b + 2 * (a * c + b * c) for a, b, c in process(text))


def part_2(text):
    return sum(a + a + b + b + a * b * c for a, b, c in process(text))


def _read_input() -> str:
    return (Path(__file__).with_name("test_data") / "day_02.txt").read_text()


def test_part_1_input():
    assert part_1(_read_input()) == 1606483

def test_part_1_example_1():
    assert part_1('2x3x4') == 58

def test_part_1_example_2():
    assert part_1('1x1x10') == 43

def test_part_2_input():
    assert part_2(_read_input()) == 3842356

def test_part_2_example_1():
    assert part_2('2x3x4') == 34

def test_part_2_example_2():
    assert part_2('1x1x10') == 14

File: rust/year_2015/src/day_02.rs
use itertools::Itertools as _;

pub fn part_1(input: &str) -> usize {
    parse_input(input)
        .map(|[l, w, h]| {
            let side_areas = [l * w, w * h, h * l];
            let surface_area: usize = side_areas.iter().map(|area| 2 * area).sum();
            let min_side_area = side_areas.into_iter().min().unwrap();
            surface_area + min_side_area
        })
        .sum()
}

pub fn part_2(input: &str) -> usize {
    parse_input(input)
        .map(|[l, w, h]| {
            let perimeters = [2 * (l + w), 2 * (w + h), 2 * (h + l)];
            let min_perimeter = perimeters.into_iter().min().unwrap();
            let volume = l * w * h;
            min_perimeter + volume
        })
        .sum()
}

fn parse_input(input: &str) -> impl Iterator<Item = [usize; 3]> {
    input.lines().map(|line| {
        line.split('x')
            .map(|n| n.parse().unwrap())
            .collect_array()
            .unwrap()
    })
}

#[cfg(test)]
mod test {
    use test_case::test_case;

    const INPUT: &str = include_str!("../test_data/day_02.txt");

    const EXAMPLE_1: &str = "2x3x4";

    const EXAMPLE_2: &str = "1x1x10";

    #[test_case(EXAMPLE_1 => 58)]
    #[test_case(EXAMPLE_2 => 43)]
    #[test_case(INPUT => 1606483)]
    fn part_1(input: &str) -> usize {
        super::part_1(input)
    }

    #[test_case(EXAMPLE_1 => 34)]
    #[test_case(EXAMPLE_2 => 14)]
    #[test_case(INPUT => 3842356)]
    fn part_2(input: &str) -> usize {
        super::part_2(input)
    }
}
